store/s3storeKMS.go

```diff
@@ -0,0 +1,478 @@
+	obj, ok, err := s.readObjectById(id)
+	if err != nil {
+		return []ChangeEvent{}, err
+	}
+
+	if !ok {
+		return []ChangeEvent{}, ErrSecretNotFound
+	}
+
+	events := []ChangeEvent{}
+
+	for ix, secretVersion := range obj.Values {
+		events = append(events, ChangeEvent{
+			Type:    getChangeType(ix),
+			Time:    secretVersion.Created,
+			User:    secretVersion.CreatedBy,
+			Version: secretVersion.Version,
+		})
+	}
+
+	// Sort events by version
+	sort.Slice(events, func(i, j int) bool {
+		return events[i].Version < events[j].Version
+	})
+	return events, nil
+}
+
+func (s *S3KMSStore) Delete(id SecretId) error {
+	index, err := s.readLatest(id.Service)
+	if err != nil {
+		return err
+	}
+
+	if val, ok := index.Latest[id.Key]; ok {
+		if val.KMSAlias != s.KMSKey() {
+			return errors.New(fmt.Sprintf("You're attempting to delete a Chamber key with a different KMS Key to the one it currently uses. Please use the current KMS Key (%s).", val.KMSAlias))
+		}
+	}
+
+	if _, ok := index.Latest[id.Key]; ok {
+		delete(index.Latest, id.Key)
+	}
+
+	if err := s.deleteObjectById(id); err != nil {
+		return err
+	}
+
+	return s.writeLatest(id.Service, index)
+}
+
+// getCurrentUser uses the STS API to get the current caller identity,
+// so that secret value changes can be correctly attributed to the right
+// aws user/role
+func (s *S3KMSStore) getCurrentUser() (string, error) {
+	resp, err := s.stsSvc.GetCallerIdentity(&sts.GetCallerIdentityInput{})
+	if err != nil {
+		return "", err
+	}
+
+	return *resp.Arn, nil
+}
+
+func (s *S3KMSStore) deleteObjectById(id SecretId) error {
+	path := getObjectPath(id)
+	return s.deleteObject(path)
+}
+
+func (s *S3KMSStore) deleteObject(path string) error {
+	deleteObjectInput := &s3.DeleteObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(path),
+	}
+
+	_, err := s.svc.DeleteObject(deleteObjectInput)
+	return err
+}
+
+func (s *S3KMSStore) readObject(path string) (secretObject, bool, error) {
+	getObjectInput := &s3.GetObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(path),
+	}
+
+	resp, err := s.svc.GetObject(getObjectInput)
+	if err != nil {
+		// handle aws errors
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchBucket:
+				return secretObject{}, false, err
+			case s3.ErrCodeNoSuchKey:
+				return secretObject{}, false, nil
+			default:
+				return secretObject{}, false, err
+			}
+		}
+		// generic errors
+		return secretObject{}, false, err
+	}
+
+	raw, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return secretObject{}, false, err
+	}
+
+	var obj secretObject
+	if err := json.Unmarshal(raw, &obj); err != nil {
+		return secretObject{}, false, err
+	}
+
+	return obj, true, nil
+
+}
+
+func (s *S3KMSStore) readObjectById(id SecretId) (secretObject, bool, error) {
+	path := getObjectPath(id)
+	return s.readObject(path)
+}
+
+func (s *S3KMSStore) puts3raw(path string, contents []byte) error {
```

Comment:
puts-3-raw? Oh, put-S3-raw. How about camelizing it? `putS3Raw`

Reply:
This is just copied from S3 Store. I didn't write it.